demux/demux.cpp

```diff
@@ -0,0 +1,229 @@
+ * interruption) however caused and on any theory of liability, whether
+ * in contract, strict liability, or tort (including negligence or
+ * otherwise) arising in any way out of the use of this software, even
+ * if advised of the possibility of such damage.
+ *
+ */
+
+
+#include "config.h"
+#include <iostream>
+#include <unirec/unirec.h>
+#include <libtrap/trap.h>
+#include <signal.h>
+#include <getopt.h>
+
+#include "fields.h"
+
+using namespace std;
+
+typedef struct meta_info_s {
+    uint16_t messageID; //1 - normal data, 2 - hello message (unirec fmt changed)
+    uint16_t interfaceID;
+    uint8_t data_fmt;
+    char payload [0];
+} meta_info_t;
+
+int HEADER_SIZE = 5;
+int exit_value = 0;
+static ur_template_t **out_templates = NULL; //UniRec output templates
+static int n_outputs = 0;
+trap_ctx_t *ctx = NULL;
+int stop = 0;
+int ret = 0;
+const void *data_nemea_output = NULL;
+uint16_t memory_received = 0;
+int verbose = 0;
+
+// Struct with information about module
+trap_module_info_t *module_info = NULL;
+
+#define MODULE_BASIC_INFO(BASIC) \
+  BASIC("demux", "This module split united input to more outputs", 1, -1)
+#define MODULE_PARAMS(PARAM) \
+PARAM('n', "link_count", "Sets count of output links. Must correspond to parameter -i (trap).", required_argument, "int32")
+
+
+int main (int argc, char ** argv)
+{
+   //int ret = 0;
+   INIT_MODULE_INFO_STRUCT(MODULE_BASIC_INFO, MODULE_PARAMS)
+   trap_ifc_spec_t ifc_spec;
+   ret = trap_parse_params(&argc, argv, &ifc_spec);
+   if (ret != TRAP_E_OK) {
+      if (ret == TRAP_E_HELP) { // "-h" was found
+         trap_print_help(module_info);
+         return 0;
+      }
+      ///fprintf(stderr, "ERROR in parsing of parameters for TRAP: %s\n", trap_last_error_msg);
+      cerr << "ERROR in parsing of parameters for TRAP: " << trap_last_error_msg << endl;
+      return 1;
+   }
+
+   // Parse remaining parameters and get configuration
+   signed char opt;
+   while ((opt = TRAP_GETOPT(argc, argv, module_getopt_string, long_options)) != -1) {
+      switch (opt) {
+      case 'n':
+         n_outputs = atoi(optarg);
+         break;
+      default:
+         //fprintf(stderr, "Error: Invalid arguments.\n");
+         cerr <<  "Error: Invalid arguments." << endl;
+         FREE_MODULE_INFO_STRUCT(MODULE_BASIC_INFO, MODULE_PARAMS)
+         return 1;
+      }
+   }
+
+   verbose = trap_get_verbose_level();
+   if (verbose >= 0) {
+      //printf("Verbosity level: %i\n", trap_get_verbose_level());
+      cout << "Verbosity level: " <<  trap_get_verbose_level();
+   }
+
+   if (verbose >= 0) {
+      //printf("Number of outputs: %i\n", n_outputs);
+      cout << "Number of outputs: " << n_outputs << endl;
+   }
+   if (n_outputs < 1) {
+      //fprintf(stderr, "Error: Number of input interfaces must be positive integer.\n");
+      cerr << "Error: Number of output interfaces must be positive integer." << endl;
+      FREE_MODULE_INFO_STRUCT(MODULE_BASIC_INFO, MODULE_PARAMS)
```

Comment:
It would be better to put this frequent code to the end and use `goto` as it is later in code.

The advantage is easier maintenance of the source code where we have all necessary cleanings on (ideally) one place.